python/ray/util/ml_utils/checkpoint_manager.py

```diff
@@ -0,0 +1,433 @@
+            checkpoints will be kept.
+        checkpoint_score_order (str). Either "max" or "min".
+            If "max", then checkpoints with highest values of
+            ``checkpoint_score_attribute`` will be kept.
+            If "min", then checkpoints with lowest values of
+            ``checkpoint_score_attribute`` will be kept.
+    """
+
+    num_to_keep: Optional[int] = None
+    checkpoint_score_attribute: Optional[str] = None
+    checkpoint_score_order: str = MAX
+
+    def __post_init__(self):
+        if self.num_to_keep is not None and self.num_to_keep < 0:
+            raise ValueError(
+                f"Received invalid num_to_keep: "
+                f"{self.num_to_keep}. "
+                f"Must be None or non-negative integer."
+            )
+        if self.checkpoint_score_order not in (MAX, MIN):
+            raise ValueError(
+                f"checkpoint_score_order must be either " f'"{MAX}" or "{MIN}".'
+            )
+
+
+@DeveloperAPI
+class CheckpointManager:
+    """Common checkpoint management and bookkeeping class for Ray Train and Tune.
+
+    This class acts as the common core for checkpoint bookkeeping in Ray ML libraries.
+    On a high level, this manager keeps a reference to all stored checkpoints
+    (both in-memory and on-disk checkpoints). For on-disk checkpoints, it
+    keeps a configured number of checkpoints according to specified metrics.
+
+    The manager supports lazy data writing by utilizing the
+    ``TrackedCheckpoint.commit()`` API, which is only invoked if the checkpoint
+    should be persisted to disk.
+
+    Args:
+        checkpoint_strategy: Checkpoint strategy defining how many and which
+            checkpoints to keep.
+        latest_checkpoint_id: First checkpoint ID to use (e.g. in case we
+            continue training an existing experiment).
+        delete_fn: Function that takes a TrackedCheckpoint and deletes it from disk
+            or memory upon request.
+
+    """
+
+    # If memory checkpoints should be persisted
+    _persist_memory_checkpoints: bool = False
+
+    def __init__(
+        self,
+        checkpoint_strategy: CheckpointStrategy,
+        latest_checkpoint_id: int = 0,
+        delete_fn: Optional[Callable[["TrackedCheckpoint"], None]] = None,
+    ):
+        self._checkpoint_strategy = checkpoint_strategy or CheckpointStrategy()
+
+        # Incremental unique checkpoint ID of this run.
+        self._latest_checkpoint_id = latest_checkpoint_id
+
+        # Used for keeping top K checkpoints.
+        self._top_persisted_checkpoints: List[_HeapCheckpointWrapper] = []
+
+        # Best checkpoint altogether.
+        # Used for exposing best_checkpoint_path.
+        self._best_persisted_checkpoint: Optional[TrackedCheckpoint] = None
+        self._latest_persisted_checkpoint: Optional[TrackedCheckpoint] = None
+        self._latest_memory_checkpoint: Optional[TrackedCheckpoint] = None
+
+        # Checkpoints that are not immediately removed
```

Comment:
Could you add a little bit more info on under what cases we need to track `checkpoints_to_clean_up` set separately?

Reply:
Added a comment